### PR TITLE
Issue 5906 - Under high pressure on entry cache, a bind on an entry m…

### PR DESCRIPTION
**Pull request by tbordaz**
…anaging retry count may trigger entry cache flush

Bug description:
	Under high pressure on the entry cache, there is a good chance
	that an unlocked entry get removed from the cache.
	During a bind, if the password policy of the target entry
	requires an update of the entry (for example
	'passwordRetryCount', the target entry may get out of the
	cache.
	In such case it will trigger entry cache flush (revert_cache)
	that is CPU intensive and that blocks others operations
	accessing the cache.

Fix description:
	If the target entry has the lockout policy, then lock
	the entry in the cache

relates: #5906

Reviewed by:

### CODE REVIEW

File: ldap/servers/slapd/back-ldbm/ldbm_bind.c [hunk 1]
```diff
@@ -63,13 +64,33 @@ ldbm_back_bind(Slapi_PBlock *pb)
...
+        passwdPolicy *pwpolicy = NULL;
+
+        /* in case of lockout password policy we may need to update
+         * the target entry (when sending the result via update_pw_retry).
+         * In such case we need lock it in the cache
+         */
+        slapi_pblock_get(pb, SLAPI_TARGET_SDN, &sdn);
+        dn = slapi_sdn_get_dn(sdn);
+        pwpolicy = new_passwdPolicy(pb, dn);
+        if (pwpolicy && (pwpolicy->pw_lockout == 1)) {
```
Comment (Firstyear) [inline ldap/servers/slapd/back-ldbm/ldbm_bind.c:79]: Doesn't this mean that these entries will be permanently pinned in the cache? That could cause the cache to thrash pretty hard in some setups.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: ldap/servers/slapd/back-ldbm/ldbm_bind.c (hunk 1 above)
Comment (tbordaz) [reply ldap/servers/slapd/back-ldbm/ldbm_bind.c:79]: You are correct, find_entry2modify pins the entry into the cache. This is the reason why the fix calls cache_unlock_entry if entrylocked is set.

Comment (Firstyear) [reply ldap/servers/slapd/back-ldbm/ldbm_bind.c:79]: Yes, but my point is there is no condition to *unpin* an entry. This means if the cache has 4000 slots, there are 3500 users, then only 500 cache slots will be usable for anything else. This problem is worse if the number of users exceeds the cache size since then they won't be cached at all I would expect.

Comment (progier389) [reply ldap/servers/slapd/back-ldbm/ldbm_bind.c:79]: Sorry but I do not understand your point:
  If I understand rightly this fix, the entry is pinned when getting it then unpinned (with cache_unlock_entry) once ldap_send_result (which may use the entry) is completed. So the maximum number of entries pinned by this fix is limited to the maximum number of worker thread (and not the number of users)

**Comment by progier389** [comment]
But I also have a concern: send_ldap_result may block until iotimeout in case of problem 
and I am afraid that the user will not being able to bind or modify its entry during that time ...

**Comment by tbordaz** [comment]
> But I also have a concern: send_ldap_result may block until iotimeout in case of problem and I am afraid that the user will not being able to bind or modify its entry during that time ...

Do you mean, there are two binds in parallel on the same entry with lockout policy. The first bind acquires the lock on the entry, fails (for any reason like wrong password), sends Invalid_credential that hangs until ioblocktimeout. Then the second bind can not proceed until ioblocktimeout ends. Am I getting your point ?
I fully agree with that scenario. So the fix should unlock the entry (if locked) before sending result. What do you think ?

**Comment by progier389** [comment]
Yes, that is the scenario that I was thinking of, 
  Your solution will fix that concern. The test should be within send_ldap_result_ext
    (after the password policy handling but before sending the first ber)

**Comment by tbordaz** [comment]
Well I have to apologize on this one because I think the reported issue does not exist.
Indeed the backend callback upload the entry in the entry cache and then when returning the result it check the password policy of the entry and may update it. Checking is done via internal search and update by internal update. If the entry gets out of the cache before checking, it is just reloaded by internal search. If the entry gets out of the cache between checking and update, it is just reloaded by internal modify.
In short I do not see how internal modify could fail (Retry count exceeded) because the entry being unlockable. Each steps backend bind, password  policy check and update have its own mechanism to get the entry and lock it if necessary.

The only benefit of this fix would be to lock the entry since the beginning of the bind (benefit of performance) but the original ticket was for robustness and IIUC I was wrong.

I propose to close that ticket. What do you think ?

**Comment by progier389** [comment]
Feel free to close this PR:
The proposed fix would need a big rework and IMHO it is no more a priority as it will not solve the customer issue that we are seeing.
And as you said the issue description is indeed wrong:
( `the target entry may get out of the cache. In such case it will trigger entry cache flush ` is  wrong )

**Comment by tbordaz** [comment]
Thanks @progier389 , @Firstyear for the reviews. I am closing this PR/ticket as the claimed issue is wrong
